### PR TITLE
MDT-69, MDT-96 Added createAndInviteUser method and Message Gateway API

### PR DESCRIPTION
**Pull request by TestaDiRapa**

### CODE REVIEW

File: test/apis/user-from-existing-patient.ts [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+        description: "London",
+        telecoms: [
+          new Telecom({
+            telecomType: "email",
+            telecomNumber: email
+          })
+        ]
+      })]
+    });
+    await userFromPatient(newPatient);
```
Comment (lisatr2812) [inline test/apis/user-from-existing-patient.ts:83]: Add some asssertions to be sure the user is properly created

File: test/apis/user-from-existing-patient.ts [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+      firstName: "Marc"
+    });
+
+    let error = undefined;
+    try {
+      await userFromPatient(newPatient);
+    } catch (e) {
+      error = e;
+    }
+    assert(!!error);
```
Comment (lisatr2812) [inline test/apis/user-from-existing-patient.ts:111]: Assert the expected error message

File: test/apis/user-from-existing-patient.ts [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+        lastName: "Specter"
+      });
+
+      let error = undefined;
+      try {
+        await userFromPatient(newPatient);
+      } catch (e) {
+        error = e;
+      }
+      assert(!!error);
```
Comment (lisatr2812) [inline test/apis/user-from-existing-patient.ts:97]: Assert the expected error message

File: test/apis/user-from-existing-patient.ts [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+      lastName: "Specter"
+    });
+
+    let error = undefined;
+    try {
+      await userFromPatient(newPatient);
+    } catch (e) {
+      error = e;
+    }
+    assert(!!error);
```
Comment (lisatr2812) [inline test/apis/user-from-existing-patient.ts:126]: Assert the expected error message

File: test/apis/user-from-existing-patient.ts [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+    try {
+      await medtechApi.userApi.createAndInviteUser(newPatient, new ICureRegistrationEmail(
+        hcp,
+        "test",
+        "iCure",
+        newPatient));
+    } catch (e) {
+      error = e;
+    }
+    assert(!!error);
```
Comment (lisatr2812) [inline test/apis/user-from-existing-patient.ts:155]: Assert the expected error message

**Review by lisatr2812 — CHANGES_REQUESTED** [review]

File: test/apis/user-from-existing-patient.ts (hunk 1 above)
Comment (TestaDiRapa) [reply test/apis/user-from-existing-patient.ts:83]: Added in the userFromPatient method, as it is shared by several tests

**Review by lisatr2812 — APPROVED** [review]

File: src/apis/impl/UserApiImpl.ts [hunk 6]
```diff
@@ -17,27 +18,95 @@ import {PaginatedListMapper} from "../../mappers/paginatedList";
...
   }
 
   async checkTokenValidity(userId: string, token: string): Promise<boolean> {
     return this.userApi.checkTokenValidity(userId, token)
   }
 
+  async createAndInviteUser(patient: Patient, messageFactory: SMSMessageFactory | EmailMessageFactory, tokenDuration = 48 * 60 * 60): Promise<User> {
+    // Checks that the Patient has all the required information
+    if (!patient.id) throw new Error("Patient does not have a valid id")
+    if (!patient.firstName) throw new Error("No first name provided in Patient");
```
Comment (lisatr2812) [inline src/apis/impl/UserApiImpl.ts:58]: We don't need those two checks : As a patient needs absolutely to have a firstName and a lastName to be created in the kraken, having patient without a firstname and a lastname should not happen. 
However, we can make sure they retrieve a real patient by : 
- checking there is a revision on this patient;
- calling a getPatient request checking the patient exists for real;

**Review by lisatr2812 — CHANGES_REQUESTED** [review]

**Review by lisatr2812 — APPROVED** [review]

**Pull request merged**